contributors/design-proposals/scheduling/pod-preemption.md

```diff
@@ -106,59 +107,299 @@ Now, assume everything in the above example, but the best effort pod has priorit
+before the scheduler considers scheduling P again. In such a case, it is likely 
+that when all the victims exit, Pod P won't fit on Node N anymore. So, scheduler
+will have to preempt other Pods on Node N or another Node so that P can be 
+scheduled. This scenario might be repeated again for the second and subsequent 
+rounds of preemption, and P might not get scheduled for a while. This scenario 
+can cause problems in various clusters, but is particularly problematic in 
+clusters with a high Pod creation rate.
 
-1. When a node becomes available, scheduler binds the pending pod to the node. The node may or may not be the same as "`NominatedNodeName`". Scheduler sets the "`NodeName`" field of PodSpec, but it does not clear "`NominatedNodeName`". See example 2 to find our reasons.
 
-### Example 1
+### Solution
 
-![image](images/preemption_1.png)
+#### Changes to the data structures
 
--  There is only 1 node (other than the master) in the cluster. The node capacity is 10 units of resources.
--  There are two pods, A and B, running on the node. Both have priority 100 and each use 5 units of resources. Pod A has 60 seconds of graceful termination period and pod B has 30 seconds.
--  Scheduler has two pods, C and D, in its queue. Pod C needs 10 units of resources and its priority is 1000. Pod D needs 2 units of resources and its priority is 50.
--  Given that pod C's priority is 1000, scheduler preempts both of pods A and B and sets the future node name of C to Node 1. Pod D cannot be scheduled anywhere.
--  After 30 seconds (or less) pod B terminates and 5 units of resources become available. Scheduler looks at pod C, but it cannot be bound yet. Pod C can still be scheduled on the node, so its future node remains the same. Scheduler tries to schedule pod D, but since pod C is ahead of pod D in the queue, scheduler assumes that it is bound to Node 1 when it evaluates feasibility of pod D. With this assumption, scheduler determines that the node does not have enough resources for pod D.
--  After 60 seconds (or less) pod A also terminates and scheduler schedules pod C on the node. Scheduler then looks at pod D, but it cannot be scheduled.
+1.  Scheduler pending queue is changed from a FIFO to a priority queue (heap). 
+The head of the queue will always be the highest priority pending pod.
+1.  A new list is added to hold unschedulable pods.
 
-### Example 2
 
-![image](images/preemption_2.png)
+#### New Scheduler Algorithm
+
+1.  Pick the head of the pending queue (highest priority pending pod).
+1.  Try to schedule the pod.
+1.  If the pod is schedulable, assume and bind it.
+1.  If the pod is not schedulable, run preemption for the pod.
+1.  Move the pod to the list of unschedulable pods.
+1.  If a node was chosen to preempt pods, set the node name as an annotation with
+ the "scheduler.kubernetes.io/nominated-node-name" key to the pod. This key is referred to as
+ "NominatedNodeName" in this doc for brevity.
```

Comment:
Can you add "When this field is set, scheduler knows that the pod is destined to run on the given node and takes it into account when making scheduling decisions for other pods" which was from the delete paragraph. It is useful to say what the point of it is, in case the reader does not figure it out from reading the algorithm.

Reply:
Sure. I added that back. PTAL.